### ISSUE TITLE
Need help with slightly complex networking

### ISSUE THREAD

**Issue by vutsalsinghal** [NginxProxyManager/nginx-proxy-manager#2074]
Hi There,

I've 2 machines: 1 on VPS (AWS) and another raspberrypi (rpi) within my home network. I use route53 as my DNS manager. I've couple of services running on both cloud vm and local rpi. 

So far I've been using ssh tunnel to expose rpi services on cloud-vm and then standalone nginx (with lets encrypt) on cloud-vm to expose all services (cloud and local) to the internet. Everything is working as expected...no concerns so far. 

Now what I want to do is, instead of opening multiple ssh tunnel ports between rpi and cloud vm, I want to use nginx-proxy-manager (npm) on rpi! So essentially there will be only 1 tunnel (port 80 for example) and npm will expose all rpi services to cloud-vm and then standalone nginx on cloud-vm will expose everything to internet.

Just for a mental pic, here is an example request path for [portainer](https://www.portainer.io/) container on rpi:
`rpi-portainer.domain.com --> cloud-vm nginx --> ssh-tunnel --> npm on rpi --> portainer container on rpi`

Things I've done so far:

1. I've created route53 entry `rpi-portainer.domain.com --> cloud-vm IP`
2. SSH tunnel running on rpi
```
#!/usr/bin/bash

ssh -i mykey.pem user@ip -R localhost:9000:localhost:80 -R localhost:9001:localhost:443 -R localhost:9002:localhost:81

```
3. standalone nginx config on cloud-vm
```
server {
    listen 80;
    listen [::]:80;
    server_name rpi-portainer.domain.com;
    return 301 https://$server_name$request_uri;
}

server {
    server_name rpi-portainer.domain.com;
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    include includes/wildcard-certs; # points to letsencrypt certs

    location / {
         proxy_pass http://localhost:9001; # Refer step 2 above. Rpi 443 is mapped to 9001 on cloud-vm.
    }
}
```
4. Docker-compose Portainer 
```
portainer:
    image: portainer/portainer-ce
    container_name: portainer_container
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./portainer:/data
    ports:
      - 16001:9000
    environment:
      - NODE_ENV=production
      - UID=1000
      - GID=1000
    restart: unless-stopped
```
5. Created ssl cert via npm UI (just in case).
6. On the npm ui, I've created a proxy host with `Domain name = rpi-portainer.domain.com` and `Forward Hostname / IP = local rpi IP` `Port = 16001`. I've tried both schemes http/https (https with the certs created on step 5).

Now when i try to load rpi-portainer.domain.com in the browser, I'm seeing the npm "Congratulations" (If you're seeing this site then you're trying to access a host that isn't set up yet) page :(

I would greatly appreciate any feedback/solution

**Comment by vutsalsinghal** [comment]
I made small change above... I switched to "host" network and when I login to the npm container, I can reach the portainer service via rpi-ip as well as docker-bridge ip (172.17.0.1). Tried both the IPs on the npm ui but still getting the same error.

**Comment by the1ts** [comment]
This particular problem is this I think a configuration issue in the nginx config in AWS. When you proxy_pass, Nginx by default strips the original host header from the request, see [here](https://docs.nginx.com/nginx/admin-guide/web-server/reverse-proxy/) in the Passing Request Headers section. Nginx will by default remove the original host header and add the host header of $proxy_host which in your above config will be localhost, therefore NPM cannot know to actually send it to rpi-portainer.domain.com. You need to add the correct header back in as NPM does for you.

In short, put the below section above the proxy_pass line in the AWS vm's Nginx config. This below, is taken directly from the NPM proxy.conf.
```
add_header       X-Served-By $host;
proxy_set_header Host $host;
proxy_set_header X-Forwarded-Scheme $scheme;
proxy_set_header X-Forwarded-Proto  $scheme;
proxy_set_header X-Forwarded-For    $remote_addr;
proxy_set_header X-Real-IP          $remote_addr;
```

This  will fix the issue of NPM not being able to know what server_name to look for in its config and showing the default_host config.
It also looks like you are having to manually copy, the NPM created wildcard certs to AWS.
All these issues will be fixed by putting NPM up front in the connection and doing simple NPM proxying to all services.
As to many or even one SSH tunnel, I use tailscale for this sort of home and cloud interconnects and for single user its free.

**Comment by vutsalsinghal** [comment]
@the1ts  you are right! I figured it out couple of hours ago. I actually got a hint when I was writing the detailed query above 😄 
thanks a lot for your input. Appreciate it!

I've used tailscale but didnt get the performance I get with simple ssh tunnel. I can stream 1080p video without any issue via jellyfin container running on my rpi. Didnt try the paid version of trailscale though. I did not do full performance analysis/comparison of the two. Its just my opinion/experience.
